File: src/ios/ScannerSettings.swift

class ScannerSettings: CustomDebugStringConvertible {
  var debugDescription: String {
    var des: String = "\(type(of: self)) {"
    for child in Mirror(reflecting: self).children {
      if let propName = child.label {
        des += "\n\t\(propName): \(child.value)"
      }
    }

    return des + "\n}"
  }

  // on android lines are much thinner, this evens out the look and feel
  private static let LINE_THICKNESS_ADJUSTMENT = -3
  // same with corner radius, way less pronounced on android
  private static var CORNER_RADIUS_ADJUSTMENT = -5

  public private(set) var barcodeFormats: Int = 1234
  public private(set) var aspectRatio: String = "1:1"
  public private(set) var aspectRatioF: Float
  public private(set) var detectorSize: Double = 0.5
  public private(set) var drawFocusRect: Bool = true
  public private(set) var focusRectColor: String = "#FFFFFF"
  public private(set) var focusRectUIColor: UIColor
  public private(set) var focusRectBorderRadius: Int = 100
  public private(set) var focusRectBorderThickness: Int = 2
  public private(set) var drawFocusLine: Bool = true
  public private(set) var focusLineColor: String = "#FFFFFF"
  public private(set) var focusLineUIColor: UIColor
  public private(set) var focusLineThickness: Int = 1
  public private(set) var drawFocusBackground: Bool = true
  public private(set) var focusBackgroundColor: String = "#CCFFFFFF"
  public private(set) var focusBackgroundUIColor: UIColor
  public private(set) var beepOnSuccess: Bool = false
  public private(set) var vibrateOnSuccess: Bool = false
  public private(set) var stableThreshold: Int = 5
  public private(set) var debugOverlay: Bool = false
  public private(set) var ignoreRotatedBarcodes: Bool = false

  init(options:[String:Any]) {
    for (key, value) in options {
      switch (key) {
      case Settings.BARCODE_FORMATS:
        if let temp = Utils.getInt(input: value) {
          barcodeFormats = temp
        }
        break
      case Settings.DETECTOR_ASPECT_RATIO:
        if let temp = value as? String {
          aspectRatio = temp
        }
        break
      case Settings.DETECTOR_SIZE:
        if let temp = Utils.getDouble(input: value) {
          detectorSize = temp
        }
        break
      case Settings.DRAW_FOCUS_RECT:
        if let temp = value as? Bool {
          drawFocusRect = temp
        }
        break
      case Settings.FOCUS_RECT_COLOR:
        if let temp = value as? String {
          focusRectColor = temp
        }
        break
      case Settings.FOCUS_RECT_BORDER_RADIUS:
        if let temp = Utils.getInt(input: value) {
          focusRectBorderRadius = max(1, temp + Self.CORNER_RADIUS_ADJUSTMENT)
        }
        break
      case Settings.FOCUS_RECT_BORDER_THICKNESS:
        if let temp = Utils.getInt(input: value) {
          focusRectBorderThickness = max(1, temp + Self.LINE_THICKNESS_ADJUSTMENT)
        }
        break
      case Settings.DRAW_FOCUS_LINE:
        if let temp = value as? Bool {
          drawFocusLine = temp
        }
        break
      case Settings.FOCUS_LINE_COLOR:
        if let temp = value as? String {
          focusLineColor = temp
        }
        break
      case Settings.FOCUS_LINE_THICKNESS:
        if let temp = Utils.getInt(input: value) {
          focusLineThickness = max(1, temp + Self.LINE_THICKNESS_ADJUSTMENT)
        }
        break
      case Settings.DRAW_FOCUS_BACKGROUND:
        if let temp = value as? Bool {
          drawFocusBackground = temp
        }
        break
      case Settings.FOCUS_BACKGROUND_COLOR:
        if let temp = value as? String {
          focusBackgroundColor = temp
        }
        break
      case Settings.BEEP_ON_SUCCESS:
        if let temp = value as? Bool {
          beepOnSuccess = temp
        }
        break
      case Settings.VIBRATE_ON_SUCCESS:
        if let temp = value as? Bool {
          vibrateOnSuccess = temp
        }
        break
      case Settings.STABLE_THRESHOLD:
        if let temp = Utils.getInt(input: value) {
          stableThreshold = temp
        }
        break
      case Settings.DEBUG_OVERLAY:
        if let temp = value as? Bool {
          debugOverlay = temp
        }
        break
      case Settings.IGNORE_ROTATED_BARCODES:
        if let temp = value as? Bool {
          ignoreRotatedBarcodes = temp
        }
        break
      default:
        print("Unknown option: \(key)")
      }
    }

    focusRectUIColor = Utils.hexStringToUIColor(hex: focusRectColor)
    focusLineUIColor = Utils.hexStringToUIColor(hex: focusLineColor)
    focusBackgroundUIColor = Utils.hexStringToUIColor(hex: focusBackgroundColor)
    aspectRatioF = Utils.getAspectRatioFromString(aspectRatioString: aspectRatio)
  }

  convenience init() {
    self.init(options: [:])
  }
}

private enum Settings {
  static let BARCODE_FORMATS: String = "barcodeFormats";
  static let DETECTOR_ASPECT_RATIO: String = "detectorAspectRatio";
  static let DETECTOR_SIZE: String = "detectorSize";
  static let DRAW_FOCUS_RECT: String = "drawFocusRect";
  static let FOCUS_RECT_COLOR: String = "focusRectColor";
  static let FOCUS_RECT_BORDER_RADIUS: String = "focusRectBorderRadius";
  static let FOCUS_RECT_BORDER_THICKNESS: String = "focusRectBorderThickness";
  static let DRAW_FOCUS_LINE: String = "drawFocusLine";
  static let FOCUS_LINE_COLOR: String = "focusLineColor";
  static let FOCUS_LINE_THICKNESS: String = "focusLineThickness";
  static let DRAW_FOCUS_BACKGROUND: String = "drawFocusBackground";
  static let FOCUS_BACKGROUND_COLOR: String = "focusBackgroundColor";
  static let BEEP_ON_SUCCESS: String = "beepOnSuccess";
  static let VIBRATE_ON_SUCCESS: String = "vibrateOnSuccess";
  static let STABLE_THRESHOLD: String = "stableThreshold";
  static let DEBUG_OVERLAY: String = "debugOverlay";
  static let IGNORE_ROTATED_BARCODES: String = "ignoreRotatedBarcodes";
}

File: src/android/com/biso/cordova/plugins/mlkit/barcode/scanner/ScannerSettings.java
package com.biso.cordova.plugins.mlkit.barcode.scanner;

import static com.biso.cordova.plugins.mlkit.barcode.scanner.ScannerSettings.Settings.*;
import static com.biso.cordova.plugins.mlkit.barcode.scanner.Utils.getAspectRatioFromString;

import android.os.Parcel;
import android.os.Parcelable;
import android.util.Log;
import java.util.Arrays;
import java.util.Iterator;
import java.util.Objects;
import java.util.Optional;
import org.json.JSONObject;

public class ScannerSettings implements Parcelable {
  private int barcodeFormats = 1234;
  private String aspectRatio = "1:1";
  private float aspectRatioF = 1;
  private double detectorSize = 0.5;
  private boolean drawFocusRect = true;
  private String focusRectColor = "#FFFFFF";
  private int focusRectBorderRadius = 100;
  private int focusRectBorderThickness = 2;
  private boolean drawFocusLine = true;
  private String focusLineColor = "#FFFFFF";
  private int focusLineThickness = 1;
  private boolean drawFocusBackground = true;
  private String focusBackgroundColor = "#CCFFFFFF";
  private boolean beepOnSuccess = false;
  private boolean vibrateOnSuccess = false;
  private int stableThreshold = 5;
  private boolean debugOverlay = false;
  private boolean ignoreRotatedBarcodes = false;

  public ScannerSettings(JSONObject settings) {
    Iterator<String> keys = settings.keys();

    while(keys.hasNext()) {
      String key = keys.next();

      Optional<Settings> setting = Settings.get(key);
      if(setting.isPresent()) {
        switch (setting.get()) {
          case BARCODE_FORMATS:
            barcodeFormats = settings.optInt(BARCODE_FORMATS.value(), getBarcodeFormats());
            break;
          case DETECTOR_ASPECT_RATIO:
            aspectRatio = settings.optString(DETECTOR_ASPECT_RATIO.value(), getAspectRatio());
            aspectRatioF = getAspectRatioFromString(aspectRatio);
            break;
          case DETECTOR_SIZE:
            double temp = settings.optDouble(DETECTOR_SIZE.value(), getDetectorSize());
            if (!(temp < 0 || temp > 1)) {
              detectorSize = temp;
            }
            break;
          case DRAW_FOCUS_RECT:
            drawFocusRect = settings.optBoolean(DRAW_FOCUS_RECT.value(), isDrawFocusRect());
            break;
          case FOCUS_RECT_COLOR:
            focusRectColor = settings.optString(FOCUS_RECT_COLOR.value(), getFocusRectColor());
            break;
          case FOCUS_RECT_BORDER_RADIUS:
            focusRectBorderRadius = settings.optInt(FOCUS_RECT_BORDER_RADIUS.value(),
                getFocusRectBorderRadius());
            break;
          case FOCUS_RECT_BORDER_THICKNESS:
            focusRectBorderThickness = settings.optInt(FOCUS_RECT_BORDER_THICKNESS.value(),
                getFocusRectBorderThickness());
            break;
          case DRAW_FOCUS_LINE:
            drawFocusLine = settings.optBoolean(DRAW_FOCUS_LINE.value(), isDrawFocusLine());
            break;
          case FOCUS_LINE_COLOR:
            focusLineColor = settings.optString(FOCUS_LINE_COLOR.value(), getFocusLineColor());
            break;
          case FOCUS_LINE_THICKNESS:
            focusLineThickness = settings.optInt(FOCUS_LINE_THICKNESS.value(),
                getFocusLineThickness());
            break;
          case DRAW_FOCUS_BACKGROUND:
            drawFocusBackground = settings.optBoolean(DRAW_FOCUS_BACKGROUND.value(),
                isDrawFocusBackground());
            break;
          case FOCUS_BACKGROUND_COLOR:
            focusBackgroundColor = settings.optString(FOCUS_BACKGROUND_COLOR.value(),
                getFocusBackgroundColor());
            break;
          case BEEP_ON_SUCCESS:
            beepOnSuccess = settings.optBoolean(BEEP_ON_SUCCESS.value(), isBeepOnSuccess());
            break;
          case VIBRATE_ON_SUCCESS:
            vibrateOnSuccess = settings.optBoolean(VIBRATE_ON_SUCCESS.value(),
                isVibrateOnSuccess());
            break;
          case STABLE_THRESHOLD:
            stableThreshold = settings.optInt(STABLE_THRESHOLD.value(), getStableThreshold());
            break;
          case DEBUG_OVERLAY:
            debugOverlay = settings.optBoolean(DEBUG_OVERLAY.value(), isDebugOverlay());
            break;
          case IGNORE_ROTATED_BARCODES:
            ignoreRotatedBarcodes = settings.optBoolean(IGNORE_ROTATED_BARCODES.value(),
                isIgnoreRotatedBarcodes());
            break;
          default:
            Log.e("SETTINGS", "No known setting for " + key);
            break;
        }
      } else {
        Log.e("SETTINGS", "No known setting for " + key);
      }
    }
  }

  public int getBarcodeFormats() {
    return barcodeFormats;
  }

  public String getAspectRatio() {
    return aspectRatio;
  }

  public float getAspectRatioF() {
    return aspectRatioF;
  }

  public double getDetectorSize() {
    return detectorSize;
  }

  public boolean isDrawFocusRect() {
    return drawFocusRect;
  }

  public String getFocusRectColor() {
    return focusRectColor;
  }

  public int getFocusRectBorderRadius() {
    return focusRectBorderRadius;
  }

  public int getFocusRectBorderThickness() {
    return focusRectBorderThickness;
  }

  public boolean isDrawFocusLine() {
    return drawFocusLine;
  }

  public String getFocusLineColor() {
    return focusLineColor;
  }

  public int getFocusLineThickness() {
    return focusLineThickness;
  }

  public boolean isDrawFocusBackground() {
    return drawFocusBackground;
  }

  public String getFocusBackgroundColor() {
    return focusBackgroundColor;
  }

  public boolean isBeepOnSuccess() {
    return beepOnSuccess;
  }

  public boolean isVibrateOnSuccess() {
    return vibrateOnSuccess;
  }

  public int getStableThreshold() {
    return stableThreshold;
  }

  public boolean isDebugOverlay() {
    return debugOverlay;
  }

  public boolean isIgnoreRotatedBarcodes() {
    return ignoreRotatedBarcodes;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ScannerSettings that = (ScannerSettings) o;
    return getBarcodeFormats() == that.getBarcodeFormats()
        && Float.compare(that.getAspectRatioF(), getAspectRatioF()) == 0
        && Double.compare(that.getDetectorSize(), getDetectorSize()) == 0
        && isDrawFocusRect() == that.isDrawFocusRect()
        && getFocusRectBorderRadius() == that.getFocusRectBorderRadius()
        && getFocusRectBorderThickness() == that.getFocusRectBorderThickness()
        && isDrawFocusLine() == that.isDrawFocusLine()
        && getFocusLineThickness() == that.getFocusLineThickness()
        && isDrawFocusBackground() == that.isDrawFocusBackground()
        && isBeepOnSuccess() == that.isBeepOnSuccess()
        && isVibrateOnSuccess() == that.isVibrateOnSuccess()
        && getStableThreshold() == that.getStableThreshold()
        && isDebugOverlay() == that.isDebugOverlay()
        && isIgnoreRotatedBarcodes() == that.isIgnoreRotatedBarcodes()
        && getAspectRatio().equals(that.getAspectRatio()) && getFocusRectColor().equals(
        that.getFocusRectColor())
        && getFocusLineColor().equals(that.getFocusLineColor()) && getFocusBackgroundColor().equals(
        that.getFocusBackgroundColor());
  }

  @Override
  public int hashCode() {
    return Objects.hash(getBarcodeFormats(), getAspectRatio(), getAspectRatioF(), getDetectorSize(),
        isDrawFocusRect(),
        getFocusRectColor(), getFocusRectBorderRadius(), getFocusRectBorderThickness(),
        isDrawFocusLine(),
        getFocusLineColor(), getFocusLineThickness(), isDrawFocusBackground(),
        getFocusBackgroundColor(),
        isBeepOnSuccess(),
        isVibrateOnSuccess(), getStableThreshold(), isDebugOverlay(), isIgnoreRotatedBarcodes());
  }

  @Override
  public int describeContents() {
    return hashCode();
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
    dest.writeInt(this.getBarcodeFormats());
    dest.writeString(this.getAspectRatio());
    dest.writeFloat(this.getAspectRatioF());
    dest.writeDouble(this.getDetectorSize());
    dest.writeByte(this.isDrawFocusRect() ? (byte) 1 : (byte) 0);
    dest.writeString(this.getFocusRectColor());
    dest.writeInt(this.getFocusRectBorderRadius());
    dest.writeInt(this.getFocusRectBorderThickness());
    dest.writeByte(this.isDrawFocusLine() ? (byte) 1 : (byte) 0);
    dest.writeString(this.getFocusLineColor());
    dest.writeInt(this.getFocusLineThickness());
    dest.writeByte(this.isDrawFocusBackground() ? (byte) 1 : (byte) 0);
    dest.writeString(this.getFocusBackgroundColor());
    dest.writeByte(this.isBeepOnSuccess() ? (byte) 1 : (byte) 0);
    dest.writeByte(this.isVibrateOnSuccess() ? (byte) 1 : (byte) 0);
    dest.writeInt(this.getStableThreshold());
    dest.writeByte(this.isDebugOverlay() ? (byte) 1 : (byte) 0);
    dest.writeByte(this.isIgnoreRotatedBarcodes() ? (byte) 1 : (byte) 0);
  }

  public void readFromParcel(Parcel source) {
    this.barcodeFormats = source.readInt();
    this.aspectRatio = source.readString();
    this.aspectRatioF = source.readFloat();
    this.detectorSize = source.readDouble();
    this.drawFocusRect = source.readByte() != 0;
    this.focusRectColor = source.readString();
    this.focusRectBorderRadius = source.readInt();
    this.focusRectBorderThickness = source.readInt();
    this.drawFocusLine = source.readByte() != 0;
    this.focusLineColor = source.readString();
    this.focusLineThickness = source.readInt();
    this.drawFocusBackground = source.readByte() != 0;
    this.focusBackgroundColor = source.readString();
    this.beepOnSuccess = source.readByte() != 0;
    this.vibrateOnSuccess = source.readByte() != 0;
    this.stableThreshold = source.readInt();
    this.debugOverlay = source.readByte() != 0;
    this.ignoreRotatedBarcodes = source.readByte() != 0;
  }

  protected ScannerSettings(Parcel in) {
    this.barcodeFormats = in.readInt();
    this.aspectRatio = in.readString();
    this.aspectRatioF = in.readFloat();
    this.detectorSize = in.readDouble();
    this.drawFocusRect = in.readByte() != 0;
    this.focusRectColor = in.readString();
    this.focusRectBorderRadius = in.readInt();
    this.focusRectBorderThickness = in.readInt();
    this.drawFocusLine = in.readByte() != 0;
    this.focusLineColor = in.readString();
    this.focusLineThickness = in.readInt();
    this.drawFocusBackground = in.readByte() != 0;
    this.focusBackgroundColor = in.readString();
    this.beepOnSuccess = in.readByte() != 0;
    this.vibrateOnSuccess = in.readByte() != 0;
    this.stableThreshold = in.readInt();
    this.debugOverlay = in.readByte() != 0;
    this.ignoreRotatedBarcodes = in.readByte() != 0;
  }

  public static final Creator<ScannerSettings> CREATOR = new Creator<ScannerSettings>() {
    @Override
    public ScannerSettings createFromParcel(Parcel source) {
      return new ScannerSettings(source);
    }

    @Override
    public ScannerSettings[] newArray(int size) {
      return new ScannerSettings[size];
    }
  };

  public enum Settings {

    BARCODE_FORMATS("barcodeFormats"),
    DETECTOR_ASPECT_RATIO("detectorAspectRatio"),
    DETECTOR_SIZE("detectorSize"),
    ROTATE_CAMERA("rotateCamera"),
    DRAW_FOCUS_RECT("drawFocusRect"),
    FOCUS_RECT_COLOR("focusRectColor"),
    FOCUS_RECT_BORDER_RADIUS("focusRectBorderRadius"),
    FOCUS_RECT_BORDER_THICKNESS("focusRectBorderThickness"),
    DRAW_FOCUS_LINE("drawFocusLine"),
    FOCUS_LINE_COLOR("focusLineColor"),
    FOCUS_LINE_THICKNESS("focusLineThickness"),
    DRAW_FOCUS_BACKGROUND("drawFocusBackground"),
    FOCUS_BACKGROUND_COLOR("focusBackgroundColor"),
    BEEP_ON_SUCCESS("beepOnSuccess"),
    VIBRATE_ON_SUCCESS("vibrateOnSuccess"),
    STABLE_THRESHOLD("stableThreshold"),
    DEBUG_OVERLAY("debugOverlay"),
    IGNORE_ROTATED_BARCODES("ignoreRotatedBarcodes");

    private final String option;

    Settings(String option) {
      this.option = option;
    }

    public String value() {
      return option;
    }

    public static Optional<Settings> get(String option) {
      return Arrays.stream(Settings.values())
          .filter(o -> o.option.equals(option))
          .findFirst();
    }
  }
}
